pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -507,119 +511,79 @@ class Measurements final
-                                                         probabilities.size()),
-                              d_probabilities);
-            return probabilities;
+        Kokkos::View<ComplexT *> sv = this->_statevector.getView();
+
+        // Reducing over `d_probabilities` requires too much L0 scratch memory
+        // on GPUs. If n_wires >= 20, this also requires quite a bit of memory
+        // on CPUs, so we fallback to the next implementation
+        if (n_wires < MDRANGE_NWIRES_MAX && !is_gpu_scratch_limited) {
+            using MDPolicyType_2D =
+                Kokkos::MDRangePolicy<Kokkos::Rank<2, Kokkos::Iterate::Left>>;
+            auto md_policy = MDPolicyType_2D(
+                {{0, 0}}, {{static_cast<int64_t>(all_indices.size()),
+                            static_cast<int64_t>(all_offsets.size())}});
+            Kokkos::parallel_reduce(
+                md_policy,
+                getProbsFunctor<PrecisionT, KokkosExecSpace>(
+                    sv, wires, d_all_indices, d_all_offsets),
+                d_probabilities);
         } else {
-            Kokkos::View<PrecisionT *> transposed_tensor("transposed_tensor",
-                                                         all_indices.size());
-
-            Kokkos::View<std::size_t *> d_trans_index("d_trans_index",
-                                                      all_indices.size());
-
-            const int num_trans_tensor = transposed_tensor.size();
-            const int num_sorted_ind_wires = sorted_ind_wires.size();
-
-            MDPolicyType_2D mdpolicy_2d1(
-                {{0, 0}}, {{num_trans_tensor, num_sorted_ind_wires}});
-
-            Kokkos::parallel_for(
-                "TransIndex", mdpolicy_2d1,
-                getTransposedIndexFunctor(d_sorted_ind_wires, d_trans_index,
-                                          num_sorted_ind_wires));
-
             Kokkos::parallel_for(
-                "Transpose",
-                Kokkos::RangePolicy<KokkosExecSpace>(0, num_trans_tensor),
-                getTransposedFunctor<PrecisionT>(
-                    transposed_tensor, d_probabilities, d_trans_index));
-
-            Kokkos::deep_copy(UnmanagedPrecisionHostView(probabilities.data(),
-                                                         probabilities.size()),
-                              transposed_tensor);
-
-            return probabilities;
+                all_indices.size(), KOKKOS_LAMBDA(const std::size_t i) {
+                    for (std::size_t j = 0; j < d_all_offsets.size(); j++) {
+                        const std::size_t index =
+                            d_all_indices(i) + d_all_offsets(j);
+                        const PrecisionT rsv = sv(index).real();
+                        const PrecisionT isv = sv(index).imag();
+                        d_probabilities(i) += rsv * rsv + isv * isv;
+                    }
+                });
         }
+
+        return view2vector(d_probabilities);
     }
```

Comment:
What is the scale of performance improvement for num_qubits < 20?"

Reply:
Let me gather some data on that.

Reply:
![speedup_vs_nqubits](https://github.com/user-attachments/assets/b663d37b-f22c-42c1-addf-7250ee59b5ec)